File: tests/driver/common/fpga/fpga_flag.cpp
// File: fpga_flag.cpp
// Project: fpga
// Created Date: 30/11/2022
// Author: Shun Suzuki
// -----
// Last Modified: 28/04/2023
// Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
// -----
// Copyright (c) 2022 Shun Suzuki. All rights reserved.
//

#include "autd3/driver/fpga/fpga_flag.hpp"

#include <gtest/gtest.h>

using autd3::driver::BitFlags;
using autd3::driver::FPGAControlFlags;

TEST(DriverCommonFPGAControlFlags, FPGAControlFlagsTest) {
  BitFlags flag(FPGAControlFlags::None);
  ASSERT_EQ(flag, FPGAControlFlags::None);

  flag.set(FPGAControlFlags::LegacyMode);
  ASSERT_TRUE(flag.contains(FPGAControlFlags::LegacyMode));
  ASSERT_FALSE(flag.contains(FPGAControlFlags::ForceFan));
  ASSERT_FALSE(flag.contains(FPGAControlFlags::ReadsFPGAInfo));
  ASSERT_FALSE(flag.contains(FPGAControlFlags::STMGainMode));
  ASSERT_FALSE(flag.contains(FPGAControlFlags::STMMode));

  flag.set(FPGAControlFlags::STMMode);
  ASSERT_TRUE(flag.contains(FPGAControlFlags::LegacyMode));
  ASSERT_TRUE(flag.contains(FPGAControlFlags::STMMode));
  ASSERT_FALSE(flag.contains(FPGAControlFlags::ForceFan));
  ASSERT_FALSE(flag.contains(FPGAControlFlags::ReadsFPGAInfo));
  ASSERT_FALSE(flag.contains(FPGAControlFlags::STMGainMode));

  flag.set(FPGAControlFlags::ForceFan);
  ASSERT_TRUE(flag.contains(FPGAControlFlags::LegacyMode));
  ASSERT_TRUE(flag.contains(FPGAControlFlags::STMMode));
  ASSERT_TRUE(flag.contains(FPGAControlFlags::ForceFan));
  ASSERT_FALSE(flag.contains(FPGAControlFlags::ReadsFPGAInfo));
  ASSERT_FALSE(flag.contains(FPGAControlFlags::STMGainMode));

  flag.set(FPGAControlFlags::ReadsFPGAInfo);
  ASSERT_TRUE(flag.contains(FPGAControlFlags::LegacyMode));
  ASSERT_TRUE(flag.contains(FPGAControlFlags::STMMode));
  ASSERT_TRUE(flag.contains(FPGAControlFlags::ForceFan));
  ASSERT_TRUE(flag.contains(FPGAControlFlags::ReadsFPGAInfo));
  ASSERT_FALSE(flag.contains(FPGAControlFlags::STMGainMode));

  flag.set(FPGAControlFlags::STMGainMode);
  ASSERT_TRUE(flag.contains(FPGAControlFlags::LegacyMode));
  ASSERT_TRUE(flag.contains(FPGAControlFlags::STMMode));
  ASSERT_TRUE(flag.contains(FPGAControlFlags::ForceFan));
  ASSERT_TRUE(flag.contains(FPGAControlFlags::ReadsFPGAInfo));
  ASSERT_TRUE(flag.contains(FPGAControlFlags::STMGainMode));

  flag.set(FPGAControlFlags::LegacyMode);
  flag.set(FPGAControlFlags::STMMode);
  flag.set(FPGAControlFlags::ForceFan);
  flag.set(FPGAControlFlags::ReadsFPGAInfo);
  flag.set(FPGAControlFlags::STMGainMode);
  ASSERT_TRUE(flag.contains(FPGAControlFlags::LegacyMode));
  ASSERT_TRUE(flag.contains(FPGAControlFlags::STMMode));
  ASSERT_TRUE(flag.contains(FPGAControlFlags::ForceFan));
  ASSERT_TRUE(flag.contains(FPGAControlFlags::ReadsFPGAInfo));
  ASSERT_TRUE(flag.contains(FPGAControlFlags::STMGainMode));

  flag.remove(FPGAControlFlags::LegacyMode);
  ASSERT_FALSE(flag.contains(FPGAControlFlags::LegacyMode));
  ASSERT_TRUE(flag.contains(FPGAControlFlags::STMMode));
  ASSERT_TRUE(flag.contains(FPGAControlFlags::ForceFan));
  ASSERT_TRUE(flag.contains(FPGAControlFlags::ReadsFPGAInfo));
  ASSERT_TRUE(flag.contains(FPGAControlFlags::STMGainMode));

  flag.remove(FPGAControlFlags::STMMode);
  ASSERT_FALSE(flag.contains(FPGAControlFlags::LegacyMode));
  ASSERT_FALSE(flag.contains(FPGAControlFlags::STMMode));
  ASSERT_TRUE(flag.contains(FPGAControlFlags::ForceFan));
  ASSERT_TRUE(flag.contains(FPGAControlFlags::ReadsFPGAInfo));
  ASSERT_TRUE(flag.contains(FPGAControlFlags::STMGainMode));

  flag.remove(FPGAControlFlags::ForceFan);
  ASSERT_FALSE(flag.contains(FPGAControlFlags::LegacyMode));
  ASSERT_FALSE(flag.contains(FPGAControlFlags::STMMode));
  ASSERT_FALSE(flag.contains(FPGAControlFlags::ForceFan));
  ASSERT_TRUE(flag.contains(FPGAControlFlags::ReadsFPGAInfo));
  ASSERT_TRUE(flag.contains(FPGAControlFlags::STMGainMode));

  flag.remove(FPGAControlFlags::ReadsFPGAInfo);
  ASSERT_FALSE(flag.contains(FPGAControlFlags::LegacyMode));
  ASSERT_FALSE(flag.contains(FPGAControlFlags::STMMode));
  ASSERT_FALSE(flag.contains(FPGAControlFlags::ForceFan));
  ASSERT_FALSE(flag.contains(FPGAControlFlags::ReadsFPGAInfo));
  ASSERT_TRUE(flag.contains(FPGAControlFlags::STMGainMode));

  flag.remove(FPGAControlFlags::STMGainMode);
  ASSERT_FALSE(flag.contains(FPGAControlFlags::LegacyMode));
  ASSERT_FALSE(flag.contains(FPGAControlFlags::STMMode));
  ASSERT_FALSE(flag.contains(FPGAControlFlags::ForceFan));
  ASSERT_FALSE(flag.contains(FPGAControlFlags::ReadsFPGAInfo));
  ASSERT_FALSE(flag.contains(FPGAControlFlags::STMGainMode));
  ASSERT_EQ(flag, FPGAControlFlags::None);

  flag.remove(FPGAControlFlags::LegacyMode);
  flag.remove(FPGAControlFlags::STMMode);
  flag.remove(FPGAControlFlags::ForceFan);
  flag.remove(FPGAControlFlags::ReadsFPGAInfo);
  flag.remove(FPGAControlFlags::STMGainMode);
  ASSERT_FALSE(flag.contains(FPGAControlFlags::LegacyMode));
  ASSERT_FALSE(flag.contains(FPGAControlFlags::STMMode));
  ASSERT_FALSE(flag.contains(FPGAControlFlags::ForceFan));
  ASSERT_FALSE(flag.contains(FPGAControlFlags::ReadsFPGAInfo));
  ASSERT_FALSE(flag.contains(FPGAControlFlags::STMGainMode));
  ASSERT_EQ(flag, FPGAControlFlags::None);
}

File: src/autd3-driver/src/fpga/fpga_flag.rs
/*
 * File: fpga_defined.rs
 * Project: src
 * Created Date: 02/05/2022
 * Author: Shun Suzuki
 * -----
 * Last Modified: 06/10/2023
 * Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
 * -----
 * Copyright (c) 2022-2023 Shun Suzuki. All rights reserved.
 *
 */

use std::fmt;

bitflags::bitflags! {
    #[derive(Clone, Copy)]
    #[repr(C)]
    pub struct FPGAControlFlags : u8 {
        const NONE            = 0;
        const FORCE_FAN       = 1 << 0;
        const READS_FPGA_INFO = 1 << 1;
    }
}

impl fmt::Display for FPGAControlFlags {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut flags = Vec::new();
        if self.contains(FPGAControlFlags::FORCE_FAN) {
            flags.push("FORCE_FAN")
        }
        if self.contains(FPGAControlFlags::READS_FPGA_INFO) {
            flags.push("READS_FPGA_INFO")
        }
        if self.is_empty() {
            flags.push("NONE")
        }
        write!(
            f,
            "{}",
            flags
                .iter()
                .map(|s| s.to_string())
                .collect::<Vec<_>>()
                .join(" | ")
        )
    }
}

#[cfg(test)]
mod tests {

    use super::*;

    #[test]
    fn fpga_info() {
        assert_eq!(std::mem::size_of::<FPGAControlFlags>(), 1);

        let flags = FPGAControlFlags::FORCE_FAN;

        let flagsc = flags.clone();
        assert_eq!(flagsc.bits(), flags.bits());
    }

    #[test]
    fn fmt() {
        assert_eq!(format!("{}", FPGAControlFlags::NONE), "NONE");
        assert_eq!(format!("{}", FPGAControlFlags::FORCE_FAN), "FORCE_FAN");
        assert_eq!(
            format!("{}", FPGAControlFlags::READS_FPGA_INFO),
            "READS_FPGA_INFO"
        );
        assert_eq!(
            format!(
                "{}",
                FPGAControlFlags::FORCE_FAN | FPGAControlFlags::READS_FPGA_INFO
            ),
            "FORCE_FAN | READS_FPGA_INFO"
        );
    }
}
